autotools: Filter libtools when building with dpcpp

cmacmackin opened the pull request:
Due to a [known issue](https://community.intel.com/t5/Intel-oneAPI-Data-Parallel-C/dpcpp-and-GNU-Autotools/m-p/1296985) with dpcpp, autotool-based builds that try to use it will fail; see #32863. This pull request modifies the Autotools package class to filter spurious object files out of the libtool script, as is already done to address issues with other compilers.

cmacmackin committed:
Filter libtools when building with dpcpp

Due to a [known
issue](https://community.intel.com/t5/Intel-oneAPI-Data-Parallel-C/dpcpp-and-GNU-Autotools/m-p/1296985)
with dpcpp, autotool-based builds that try to use it will fail because
they try to link against temporary files that no longer exist. This
commit filters those files out of the libtools script so that linking
can work properly.

alalazo commented:
@spackbot run pipeline

haampie commented:
I think the solution is alright, but is this still an issue with the latest dpcpp? If so, can you please ask for the status regarding

> We are working on your issue internally. We will get back to you soon.

Otherwise:

1. Please add a comment clearly explaining why this change was necessary, linking to the original issue.
2. Restrict the change to the lines starting with `predep_objects_...` to avoid false positives.

haampie reviewed (changes requested):
I think the solution is alright. Also, can you please ask for the status regarding

> We are working on your issue internally. We will get back to you soon.

on their forum / helpdesk? Otherwise we're stuck with these hacks forever...

Otherwise:

1. Please add a comment clearly explaining why this change was necessary, linking to the original issue.
2. Restrict the change to the lines starting with `predep_objects_...` to avoid false positives.

cmacmackin committed:
Make filtering of predep_objects for dpcpp more specific

It will now only filter out these files from the predep_objects
line. A comment with a reference to the issue was also added.

cmacmackin commented:
> I think the solution is alright. Also, can you please ask for the status regarding
> 
> > We are working on your issue internally. We will get back to you soon.
> 
> on their forum / helpdesk? Otherwise we're stuck with these hacks forever...

I wasn't the original person to create that thread, but I've [posted a reply](community.intel.com/t5/Intel-oneAPI-Data-Parallel-C/dpcpp-and-GNU-Autotools/m-p/1419079#M2568) asking about this.
 
> Otherwise:
> 
>     1. Please add a comment clearly explaining why this change was necessary, linking to the original issue.
> 
>     2. Restrict the change to the lines starting with `predep_objects_...` to avoid false positives.

I've just pushed a new commit with these changes.

cmacmackin committed:
Fixed overly long lines

cmacmackin committed:
Reverted incorrectly breaking a function call onto multiple lines
